### DISCUSSION TITLE
Self-hosting

### DISCUSSION THREAD

**Discussion by joebudi** [Budibase/budibase#648] [Help]
We are currently in the planning phase for creating and documenting a process that helps users self-host their Budibase app.
We will strive to make this a simple experience, but this will require input from the community. I have opened this discussion as a vehicle for collaboration, so the community and our team can work towards the best possible solution for everyone.

We have an internal meeting coming up regarding the topic. I will share notes from the meeting within this discussion.

**Comment by LinuxAndCoffee** [comment]
I would love to look at the self-hosted option. Not that I think your pricing is unreasonable. I am a pastor (formerly IT) and have limited funding available for this kind of work, even though my completed apps should help me be more organized and save time. Please let me know what information I could provide to help in your work. Thank you for this awesome tool!

P.S. - Any idea what the general requirements would be for self-hosting (i.e. - OS, RAM, database, etc.)? And any chance the final app that is made could run on a shared hosting platform?

**Reply by mjashanks** [reply]
Hi @revmwv . I can't give you exact minimum requirements just a the moment. However, I can comfortably run Budibase on my home laptop, that I bought for $500 in 2012!

Budibase is cross platform. It runs NodeJS + Koa on the server, and CouchDb as a database.

The final app that you build with budibase runs on exactly the same codebase, thus has the same requirements - I.e. you should be able to run it anywhere.

**Reply by LinuxAndCoffee** [reply]
Awesome! I am really excited about Budibase. You are all doing great work and I feel like the possibilities for what we could create with it are endless!

**Comment by osamaros** [comment]
I am too interested in self-hosting. I already have my infrastructure for it.

**Reply by mjashanks** [reply]
Sit tight - we want to get Budibase stable, with Beta testing, before we support self hosting :)

**Reply by abasse** [reply]
Any update on self hosting? Very keen to get if a try.

**Reply by joebudi** [reply]
Hey. It's not possible to self host at the moment, but we are researching the best approach. Would you be free for a chat re self hosting and how we can build a great experience?

**Reply by abasse** [reply]
Hi, happy to chat about the hosting options. I have a dev team who can also offer support if required. You can reach me at andre.basse@spiegel.de

**Reply by Wulfheart** [reply]
@joebudi Man, @abasse has an mail address from spiegel.de. They are one of the biggest German magazines.

**Comment by Wulfheart** [comment]
Does a self hosted version allow to remove the budibase logo?

**Reply by n0c0de-com** [reply]
I haven't read the licensing rules. But, you can technically do anything with Budi if you figure out the source. I have the dev hosted at https://budi.n0c0de.com
But, I would encourage giving the team as much credit as possible to help them develop the project.

Here is a tiny project for POC: https://budi.n0c0de.com/app_2ad26e3128c94dd09657c5b787144694

**Reply by mjashanks** [reply]
Hi Alex. If you are wanting to remove the Budibase logo inside your app, you can do it - regardless of self-hosting.
- Click "Master Screen"
- Click "Navigation"
- On the panel on the right, set the logo URL

![image](https://user-images.githubusercontent.com/3524181/101916417-d0552c80-3bbe-11eb-9edf-3588957ddbd2.png)

If it's the logo on the login screen, 
- Click the "Public" section
- Click "Login"
- Change the logo URL on the right-hand side.

**Comment by andrewdmaclean** [comment]
@mjashanks & @joebudi this is a nearly perfect platform for transitioning students between tools like Figma and building their own application from scratch!

Way to go on an awesome tool and looking forward to piloting with students in early 2021.

**Reply by n0c0de-com** [reply]
This is a great idea based my experience so far!

**Reply by joebudi** [reply]
That's a really interesting use case. We've had lecturers from some great universities sign-up and it would be great to help the next generation of developers.

**Reply by andrewdmaclean** [reply]
@joebudi amazing! Looking forward to discovering how the self-hosting market will work.

**Reply by joebudi** [reply]
@andrewdmaclean @mike12345567 is working on it at the moment. Would you be free for a very, very quick chat? I'd love to understand how to deliver a self-hosting experience that delights you.

**Reply by andrewdmaclean** [reply]
@joebudi absolutely 👍 I don't want to throw my email out in a public discussion, but is there a budibase email that I can connect with you at?

**Reply by joebudi** [reply]
joe at budibase dot com :-)

**Reply by andrewdmaclean** [reply]
@joebudi just sent you an email 😄 Looking forward to connecting with you soon.

**Reply by joebudi** [reply]
@andrewdmaclean self-hosting is up and running. Let me know if you need a hand getting it setup.

**Comment by supermx1** [comment]
Hello, great app. I am new to development as a whole. I use Bootstrap/WordPress a lot for web design and I have always wanted to dive into backend development. I stumbled on budibase in my extensive search for a low code platform that can help noobs like me get backend skills in. Now I am used to hosting on apache/Linux (aka Shared hosting), I have hosted countless websites on a shared hosting/vps. For people like me that don't know how to use node.js or koa or anything in that line how can the self-hosting be made easier for us? Plus it is limited to docker and digital ocean, which is not popular in my country (we mostly use cpanel based hostings).
Is there an easier way for hosting offline for testing on something like Xampp or Mamp?

**Reply by joebudi** [reply]
Hey. I come from a very similar background to you. For 10 years of my career, I used Wordpress to build websites for clients. I could never build web apps, so here we (a number of other amazing devs) are building a web app builder :-). I promise you, Digital Ocean is pretty straight forward and easier, in my opinion, than cpanel.

We're on their marketplace, so hosting takes just a few clicks:
https://marketplace.digitalocean.com/apps/budibase

**Reply by supermx1** [reply]
Okay, let me try to explain an issue as best I can. You see in my country, there is a law being pushed that all applications built to collect data or data handling applications must be hosted within the country (i.e no cloud hosting services). This has really narrowed our reach in terms of distribution. Most on-premise hosting services or just hosting companies as a whole only offer linux/apache/windows hosting services, this further narrows the path. So I am sort of asking for a feature that includes users like us that have been boxed in by law. How can we implement budibuilder on either localhost or the limited hosting services we can find

**Reply by mjashanks** [reply]
Hey @emekatoby - we also support docker-compose: https://docs.budibase.com/self-hosting/hosting-methods/docker-compose

This should allow you to run it anywhere.

**Reply by charleschoquetteevo** [reply]
@joebudi  How do I get the hosting key on the Digital Ocean Droplet ?

**Reply by mjashanks** [reply]
hey @charleschoquetteevo . This is described here: https://marketplace.digitalocean.com/apps/budibase under "Configuring Your Budibase Installation" :)

**Reply by charleschoquetteevo** [reply]
@mjashanks And what is the right command to edit hosting properties?  Do you have a video?  Thanks !

**Reply by mjashanks** [reply]
in your droplet, you should find a `hosting.properties` file. in here, there should be a `HOSTING_KEY` property for you to edit. Restart your containers afterwards.

I'm not 100% about this, might have to get someone else to weigh in here. But this worked for me, running self-hosted locally with docker-compose.

**Comment by supermx1** [comment]
Also, I see no options to export the code, so as to deploy on a server that runs on NodeJS. Is there a way to export the code locally?

**Reply by mjashanks** [reply]
Budibase includes the full stack. Apps built with budibase run on 
- NodeJS +Koa
- CouchDB
- Svelte on the front-end

When you self host Budibase, using docker, you run the full budibase stack. There is no reason that I can think of to export. Does this cover what you need?

**Reply by supermx1** [reply]
Maybe I am still unclear on how self-hosting Budibase works. Budibase application on Mac is a standalone app. You can create multiple applications. Now to host each individual application created using Budibase still confuses me. Hence why i asked for the code export option. The self-hosting option under settings only has a hosting URL and hosting key. So if i have multiple projects set up how does the software handle that in docker?

**Reply by mjashanks** [reply]
So, you build your apps in the desktop app - "Budibase Builder". You can also preview them here, to test them.

![image](https://user-images.githubusercontent.com/3524181/106741732-0093d700-6614-11eb-8b3c-49697e7657ff.png)

When you are ready to deploy your apps / push them to production, you "deploy" via the Deploy section in the desktop app.

![image](https://user-images.githubusercontent.com/3524181/106741772-130e1080-6614-11eb-9238-933ba077bfa9.png)

"Deploy App" will package up your app, and send it to your self-hosted server. Your deployed app will then be available via a URL on your self-hosted server. 

Any changes that you make in the desktop app will only be available locally (Preview) until you press deploy again.

You can deploy as many apps as you like, to the same self-host instance.

**Reply by supermx1** [reply]
Yes, I am aware of this. Now when I deploy a test app. Its opens a URL, I am guessing on Budibase servers. This is fine, when one wants to quickly preview a build. My question is one has multiple projects and you want deploy each of them in their own docker instances or droplets(self-hosted server) if I may use those terms; Is there a way to do that currently? From my very limited knowledge, I can only see one option in the setting > self-hosting section. You can only add a setting for a hosting URL and a hosting key. This works when you have a single project how do you deploy multiple projects to multiple self-hosted instances, without having to always go and change the hosting url and key each time you want to deploy to a different self-hosted service?

I don't know if I am being clear?

**Reply by mjashanks** [reply]
Ok I understand what you are asking for now :)

At the moment, you cannot set up multiple self-host instances and choose which one to deploy to (without changing the hosting URL and key each time).

This is something that we have talked about internally, but nothing planned yet. To me, it makes sense to allow this.

Like I say, you can deploy many apps to the same server. Their URL will take the form of `https://<self-hosted-url>/apps/<app-name>`. 

However, it seems that this is not what you are looking for. I'd love to know the reason why this doesn't suit your use case - I'd be helpful for us to know what you are trying to achieve.

Apologies for not answering the question!

**Comment by christianechevarria** [comment]
Hi everyone, I added a question, https://github.com/Budibase/budibase/discussions/1633, on how to access the 1-click digital ocean budibase app to see if anyone would know something about it -- maybe it'll help someone else wondering the same thing

**Reply by joebudi** [reply]
Hey @chrisgereina . I see @shogunpurple answered your question, and thanks for linking.

**Comment by slfhstr** [comment]
IMHO the biggest "use case" for self-hosting is data protection : certainty and compliance.
It's not about trust (or any lack of it).
It's about removing a layer of potential exposure.
And de-risking about any future changes to GDPR, and issues on location of data.
Using a supplier-hosted SaaS opens up questions : self-hosting closes them down.
So ability to self-host Budibase has always been a massive positive, over other no-code/no-code offerings.

**Reply by joebudi** [reply]
Hey @timconsidine . These are exactly our thoughts too. From conversations with users, data compliance is the number 1 reason for self-hosting.

**Comment by marcosfelt** [comment]
Really cool project! Is there anything that would restrict us from using budibase for a saas app?  In other words, can we use the builder internally and, then, use the self-hosted deploy functionality to serve our clients?

**Reply by joebudi** [reply]
Hey

Thanks for your kind words! Budibase is not designed for SaaS apps. So, you might find the signup process, and lack of payment solution/connection problematic when building a SaaS.

**Reply by marcosfelt** [reply]
That's perfectly fine, we really just need something to build out a PoC at this point. As long as the license is fine, we'll take a look!

**Comment by KTheMan** [comment]
It's working great for me just had to change the couchdb port back to 5984. It wasn't looking for it on 4005, despite it being the default in the hosting.properties file.

**Comment by jaspreet1313** [comment]
how to self host budibase on cloud

**Reply by shogunpurple** [reply]
https://docs.budibase.com/self-hosting/hosting-methods

**Reply by jaspreet1313** [reply]
But When i follow these steps it's not shown any hosting option and not any signup page as shown in this video {"https://youtu.be/2P4HaFkP6Zk")

**Reply by mjashanks** [reply]
We don't yet have AWS, Azure or GCP specific deployment templates.

However, you can create a Linux VM on any of those platforms, and follow the setup instructions via `docker-compose`. 

Or you can use `Kubernetes`

**Reply by jaspreet1313** [reply]
I have done with installation of docker ,docker compose for self hosting from "https://docs.budibase.com/self-hosting/hosting-methods/docker-compose".
But when i done with this I create a app and publish it .It did not give any published link  so i can use it without login .Everytime it's asking for login credential even after i published it.
please reply asap.
Thanku in advance :-)

**Reply by mjashanks** [reply]
@jaspreet1313 - if you want to access a budibase app without login, you will need to make some "Public" screens, and browse directly to the URL for those screens. 

You will also need to make sure that your public screens are only accessing tables/queries that have "Public" access. Beware that this will mean that all data in those tables will be public (if "Read" is set to Public)

**Reply by jaspreet1313** [reply]
Thanku I wil try

**Comment by kanchan1427** [comment]
i have installed the docker and docker compose,budibase cli as it was mentioned in the documentation and i am able to create an app or to use an template but after creating and making apps  public it always requires login credentials .There isn't any option for deployement. can somebody please tell me how can i deploy my app or is there anything i am doing wrong

**Reply by mjashanks** [reply]
You deploy your app using the `Publish` button:
![image](https://user-images.githubusercontent.com/3524181/153835515-f4369916-eafd-4813-9564-7c4c9a2016e9.png)

If you want your app to work without login, you will need to do this per screen. Click on the screen that you want to make public and set the `Access` to `Public`

If your app reads or writes to any tables/datasource, you will also need to give these these "Public" access

**Comment by mjashanks** [comment]
going to lock this discussion as it's gone a bit off-topic - please submit a new one if you have a specific question
